deepspeed/runtime/zenflow/engine_stage3.py

```diff
@@ -0,0 +1,748 @@
+            cuda_grad_buffer = grad_buffer.to(grad_partition.device, non_blocking=True)
+            cuda_grad_buffer.add_(
+                grad_partition.to(optimizer_z3.gradient_accumulation_dtype).view(cuda_grad_buffer.shape))
+            grad_buffer.copy_(cuda_grad_buffer, non_blocking=True)
+            # ensure grad buffer is a CUDA buffer to speed up the next few
+            # operations and so it can be used asynchronously
+            grad_buffer = cuda_grad_buffer
+
+        # offload the gradient partition if applicable
+        if optimizer_z3.offload_optimizer:
+            i, dest_offset, _ = optimizer_z3.grad_position[optimizer_z3.get_param_id(param)]
+            now_state = optimizer_z3.get_overlap_step_state()
+
+            if optimizer_z3.is_gradient_accumulation_boundary:
+                optimizer_z3.norm_for_param_grads[optimizer_z3.get_param_id(
+                    param)] = optimizer_z3._constant_buffered_norm2(grad_buffer)
+
+                if optimizer_z3._swappable_optimizer_subgroup(i):
+                    if not i in offload_fp32_gradients.keys():
+                        offload_fp32_gradients[i] = []
+                        offload_fp32_offsets[i] = []
+
+                    offload_fp32_gradients[i].append(grad_buffer.float())
+                    offload_fp32_offsets[i].append(dest_offset)
+                else:
+                    fp32_grad_tensor = optimizer_z3.fp32_partitioned_groups_flat[i].overlap_grad[now_state].narrow(
+                        0, dest_offset, grad_buffer.numel())
+                    fp32_grad_tensor.copy_(grad_buffer.float())
+
+        # free the gradient
+        if not get_accelerator().is_synchronized_device():
+            if param.grad is not None:
+                param.grad.record_stream(get_accelerator().current_stream())
+        param.grad = None
+
+    if optimizer_z3.offload_optimizer and optimizer_z3.swap_optimizer:
+        for i in offload_fp32_gradients.keys():
+            optimizer_z3.optimizer_swapper.swap_out_gradients(parameter=optimizer_z3.fp32_partitioned_groups_flat[i],
+                                                              gradient_offsets=offload_fp32_offsets[i],
+                                                              gradient_tensors=offload_fp32_gradients[i])
+    return buffers
+
+
+def disable_accelerator():
+    accelerator = get_accelerator()
+    accelerator.is_available = lambda: False
+    accelerator.device_count = lambda: 0
+    accelerator.current_device = lambda: -1
+    # Optionally mark it as initialized if needed
+    if hasattr(accelerator, "_initialized"):
+        accelerator._initialized = True
+
+
+def zenflow_optimizer_process(pipe, curr_rank, total_rank, param_groups, shared_overlap_grad_map,
+                              shared_stale_param_map):
+    os.environ["CUDA_VISIBLE_DEVICES"] = ""
+    disable_accelerator()
+
+    TOTAL_CORES = os.cpu_count()
```

Comment:
Hi @JoshWoo2003 ,
@Antlera and me had improved ZenFlow affinity setting in the following PR.  I think we should do the same for ZenFlow for stage3.  Can you integrate the affinity setting code in your PR as well?  In that way we can have unified ZenFlow affinity behavior with different stages.  Thanks!
https://github.com/deepspeedai/DeepSpeed/pull/7506